src/parser.ts

```diff
@@ -91,13 +109,40 @@ const areAddressesValid = (row: Payment): string[] => {
 const isAmountPositive = (row: Payment): string[] =>
   row.amount.isGreaterThan(0) ? [] : ["Only positive amounts possible: " + row.amount.toFixed()];
 
-const isDecimalValid = (row: Payment, tokenList: TokenMap): string[] => {
-  if (row.tokenAddress == null || row.tokenAddress === "") {
-    return [];
+const isTokenValid = (row: Payment): string[] =>
+  row.decimals === -1 && row.symbol === "SYMBOL_UNKNOWN"
+    ? [`No valid token contract with tokens was found at ${row.tokenAddress}`]
+    : [];
+
+export async function toPayment(prePayment: PrePayment, tokenInfoProvider: TokenInfoProvider): Promise<Payment> {
+  if (prePayment.tokenAddress === null) {
+    // Native asset payment.
+    return {
+      receiver: prePayment.receiver,
+      amount: prePayment.amount,
+      tokenAddress: prePayment.tokenAddress,
+      decimals: 18,
+      symbol: "ETH",
+    };
+  }
+  const tokenInfo = await tokenInfoProvider.getTokenInfo(prePayment.tokenAddress);
+  if (typeof tokenInfo !== "undefined") {
+    let decimals = tokenInfo.decimals;
+    let symbol = tokenInfo.symbol;
+    return {
+      receiver: prePayment.receiver,
+      amount: prePayment.amount,
+      tokenAddress: prePayment.tokenAddress,
+      decimals,
+      symbol,
+    };
   } else {
-    const decimals =
-      tokenList.get(utils.isAddress(row.tokenAddress) ? utils.getAddress(row.tokenAddress) : row.tokenAddress)
-        ?.decimals || row.decimals;
-    return decimals >= 0 ? [] : ["Invalid decimals: " + decimals];
+    return {
+      receiver: prePayment.receiver,
+      amount: prePayment.amount,
+      tokenAddress: prePayment.tokenAddress,
+      decimals: -1,
+      symbol: "SYMBOL_UNKNOWN",
+    };
```

Comment:
I guess this is a followup to my question about the scenario where we have the decimals but not the symbol. In which case, I might even put the token address in the symbol field... Not sure if this every happens though

Reply:
Symbol field may be undefined in the newest version. 
We dont need to put the address in the symbol because we already have the address in this object.